File: MANUAL_GITHUB/regions.py
from __future__ import annotations

import re
import unicodedata

CANADA = "CA"
PROVINCE_CODES = (
    "QC", "ON", "BC", "AB", "SK",
    "MB", "NB", "NS", "PE", "NL",
)
ALL_CANADIAN_REGIONS = (CANADA, *PROVINCE_CODES)

_REGION_ALIASES: dict[str, tuple[str, ...]] = {
    "QC": (
        "quebec", "québec",
    ),
    "ON": (
        "ontario",
    ),
    "BC": (
        "british columbia", "b.c.", "b.c", "colombie-britannique",
    ),
    "AB": (
        "alberta",
    ),
    "SK": (
        "saskatchewan",
    ),
    "MB": (
        "manitoba",
    ),
    "NB": (
        "new brunswick", "nouveau-brunswick", "n.b.", "n.-b.",
    ),
    "NS": (
        "nova scotia", "nouvelle-ecosse", "nouvelle-écosse", "n.s.", "n.-e.",
    ),
    "PE": (
        "prince edward island", "p.e.i.", "pei",
        "ile-du-prince-edouard", "île-du-prince-édouard", "i.-p.-e.",
    ),
    "NL": (
        "newfoundland and labrador", "newfoundland & labrador",
        "terre-neuve-et-labrador", "terre neuve et labrador", "n.l.", "t.-n.-l.",
    ),
}

# Releases in these families routinely contain provincial tables/series even
# when the headline is national. They should remain visible inside a province
# view in Anatole.
_PROVINCIAL_BREAKDOWN_PATTERNS = (
    "labour force survey",
    "enquete sur la population active",
    "consumer price index",
    "indice des prix a la consommation",
    "retail trade",
    "commerce de detail",
    "wholesale trade",
    "commerce de gros",
    "building permits",
    "permis de batir",
    "manufacturing sales",
    "monthly survey of manufacturing",
    "enquete mensuelle sur les industries manufacturieres",
    "payroll employment",
    "emploi salarie",
    "job vacancies",
    "postes vacants",
    "investment in building construction",
    "investissement en construction de batiments",
    "new housing price index",
    "indice des prix des logements neufs",
    "population estimates",
    "estimations demographiques",
    "gross domestic product by industry: provinces",
    "produit interieur brut par industrie : provinces",
    "provincial and territorial economic accounts",
    "comptes economiques provinciaux et territoriaux",
)


def _fold(value: str) -> str:
    normalized = unicodedata.normalize("NFKD", value.casefold())
    ascii_text = "".join(ch for ch in normalized if not unicodedata.combining(ch))
    return re.sub(r"\s+", " ", ascii_text).strip()


def explicit_regions(text: str) -> list[str]:
    folded = _fold(text)
    searchable = re.sub(r"[^a-z0-9]+", " ", folded).strip()
    padded = f" {searchable} "
    regions: list[str] = []

    for code in PROVINCE_CODES:
        aliases = _REGION_ALIASES[code]
        for alias in aliases:
            alias_folded = _fold(alias)
            alias_searchable = re.sub(
                r"[^a-z0-9]+",
                " ",
                alias_folded,
            ).strip()
            if alias_searchable and f" {alias_searchable} " in padded:
                regions.append(code)
                break

    return regions


def economic_regions(text: str) -> list[str]:
    explicit = explicit_regions(text)
    if explicit:
        return explicit

    folded = _fold(text)
    if any(pattern in folded for pattern in _PROVINCIAL_BREAKDOWN_PATTERNS):
        return list(ALL_CANADIAN_REGIONS)

    return [CANADA]


def province_region(code: str) -> list[str]:
    return [code] if code in PROVINCE_CODES else [CANADA]

File: MANUAL_GITHUB/regions.ts
import type {
  AnatoleLanguage,
} from "@/lib/preferences";

export type RegionCode =
  | "ALL"
  | "CA"
  | "QC"
  | "ON"
  | "BC"
  | "AB"
  | "SK"
  | "MB"
  | "NB"
  | "NS"
  | "PE"
  | "NL";

export const REGION_CODES: RegionCode[] = [
  "ALL",
  "CA",
  "QC",
  "ON",
  "BC",
  "AB",
  "SK",
  "MB",
  "NB",
  "NS",
  "PE",
  "NL",
];

const LABELS: Record<
  Exclude<RegionCode, "ALL">,
  { fr: string; en: string }
> = {
  CA: { fr: "Canada", en: "Canada" },
  QC: { fr: "Québec", en: "Quebec" },
  ON: { fr: "Ontario", en: "Ontario" },
  BC: { fr: "Colombie-Britannique", en: "British Columbia" },
  AB: { fr: "Alberta", en: "Alberta" },
  SK: { fr: "Saskatchewan", en: "Saskatchewan" },
  MB: { fr: "Manitoba", en: "Manitoba" },
  NB: { fr: "Nouveau-Brunswick", en: "New Brunswick" },
  NS: { fr: "Nouvelle-Écosse", en: "Nova Scotia" },
  PE: { fr: "Île-du-Prince-Édouard", en: "Prince Edward Island" },
  NL: { fr: "Terre-Neuve-et-Labrador", en: "Newfoundland and Labrador" },
};

export function regionLabel(
  code: RegionCode,
  language: AnatoleLanguage,
): string {
  if (code === "ALL") {
    return language === "fr" ? "Toutes" : "All";
  }
  return LABELS[code][language];
}

export function itemRegions(
  regions: string[] | undefined,
): string[] {
  return regions?.length ? regions : ["CA"];
}

export function matchesRegion(
  regions: string[] | undefined,
  selected: RegionCode,
): boolean {
  if (selected === "ALL") return true;
  const values = itemRegions(regions);
  if (selected === "CA") {
    return values.includes("CA");
  }
  // Province view intentionally keeps Canada-wide releases visible as the
  // shared context requested by the product.
  return values.includes(selected) || values.includes("CA");
}

export function regionSummary(
  regions: string[] | undefined,
  language: AnatoleLanguage,
): string {
  const values = itemRegions(regions);
  const provinces = values.filter((code) => code !== "CA");

  if (provinces.length >= 8) {
    return language === "fr"
      ? "Canada + provinces"
      : "Canada + provinces";
  }

  if (provinces.length === 1 && !values.includes("CA")) {
    return regionLabel(provinces[0] as RegionCode, language);
  }

  if (provinces.length > 1) {
    return provinces
      .slice(0, 2)
      .map((code) => regionLabel(code as RegionCode, language))
      .join(" · ") + (provinces.length > 2 ? " +" : "");
  }

  return regionLabel("CA", language);
}
